src/librustc/middle/infer/mod.rs

```diff
@@ -98,6 +98,9 @@ pub struct InferCtxt<'a, 'tcx: 'a> {
     normalize: bool,
 
     err_count_on_creation: usize,
+
+    // Default Type Parameter fallbacks
+    pub defaults: RefCell<FnvHashMap<Ty<'tcx>, Ty<'tcx>>>,
```

Comment:
So, I'm concerned about this map.

For one thing, I'd want to "encapsulate it more":

a. make this private
b. key it by `ty::TyVid`
c. insert new data by adding an (optional) parameter to `next_ty_var`, rather than mutating the table after the fact

But a bigger issue is that it should be versioned for transactions. As it is, we are going to create fresh type variables with defaults then (potentially) roll them back, which might clear the default. I think my preference would be to remove the map, and instead store the `Option<Ty<'tcx>>` as part of the `Bounded` variant in `type_variable::TypeVariableValue` (after all, who cares what the default was once the type is `Known`?)

I'd probably change `TypeVariableValue` to something like:

```rust
enum TypeVariableValue<'tcx> {
    Known(Ty<'tcx>),
    Bounded { relations: Vec<Relation>, default: Option<Ty<'tcx>> }
}
```

(Also, if you do it this way, the transactional problem should just take care of itself.)